src/main/scala/com/databricks/spark/avro/AvroSaver.scala

```diff
@@ -0,0 +1,205 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.databricks.spark.avro
+
+import java.sql.Timestamp
+import java.util.HashMap
+
+import scala.collection.immutable.Map
+
+import org.apache.spark.sql._
+import org.apache.spark.SparkContext._
+
+import org.apache.avro.generic.GenericData.Record
+import org.apache.avro.generic.GenericRecord
+import org.apache.avro.mapred._
+import org.apache.avro.SchemaBuilder.{RecordBuilder, FieldAssembler}
+import org.apache.avro.{SchemaBuilder, Schema}
+
+import org.apache.hadoop.io.NullWritable
+import org.apache.hadoop.mapred.JobConf
+
+
+/** 
+ * This object provides a save() method that is used to save SchemaRDD as avro file.
+ * To do this, we first convert the schema and then convert each row of the RDD to corresponding
+ * avro types. One remark worth mentioning is the structName parameter that functions have. Avro
+ * records have a name associated with them, which must be unique. Since SturctType in sparkSQL
+ * doesn't have a name associated with it, we are taking the name of the last structure field that
+ * the current structure is a child of. For example if the row at the top level had a field called
+ * "X", which happens to be a structure, we would call that structure "X". When we process original
+ * rows, they get a name "topLevelRecord".
+ */
+object AvroSaver {
+
+  private def convertTypeToAvro[T](dataType: DataType, 
+      schemaBuilder: SchemaBuilder.BaseTypeBuilder[T],
+      structName: String): T = {
+    dataType match {
+      case ByteType => schemaBuilder.bytesType()
+      case ShortType => schemaBuilder.intType()
+      case IntegerType => schemaBuilder.intType()
+      case LongType => schemaBuilder.longType()
+      case FloatType => schemaBuilder.floatType()
+      case DoubleType => schemaBuilder.doubleType()
+      case DecimalType() => schemaBuilder.stringType()
+      case StringType => schemaBuilder.stringType()
+      case BinaryType => schemaBuilder.bytesType()
+      case BooleanType => schemaBuilder.booleanType()
+      case TimestampType => schemaBuilder.longType()
+
+      case ArrayType(elementType, _) =>
+        val builder = SchemaBuilder.builder()
+        val elementSchema = convertTypeToAvro(elementType, builder, structName)
+
+        schemaBuilder.array().items(elementSchema)
+
+      case MapType(StringType, valueType, _) =>
+        val builder = SchemaBuilder.builder()
+        val valueSchema = convertTypeToAvro(valueType, builder, structName)
+
+        schemaBuilder.map().values(valueSchema)
+
+      case StructType(_) =>
+        convertStructToAvro(dataType.asInstanceOf[StructType],
```

Comment:
If you need to access the typed variable, you can do this funny syntax:

```scala
case structType @ StructType(_) =>
  convertStructToAvro(structType, schemaBuilder.record(structName))
```

However, since you're not actually using the pattern-match, it may be easier to read to just use
```scala
case structType: StructType =>
  convertStructToAvro(structType, schemaBuilder.record(structName))
```